### PR TITLE
feat: connection check retry in soef

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

The SOEF connection conducts a connection check during initial connect call by multiplexer. If this fails it stops the agent. Now we add a max retry logic to avoid agent unnecessary failing if the first connection attempt fails.

## Fixes

-

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: examples/tac_deploy/README.md [hunk 1]
```diff
@@ -95,8 +95,10 @@ grep -rl 'TAKE CARE! Circumventing controller identity check!' output_dir/ | sor
 grep -rl 'TAKE CARE! Circumventing controller identity check!' output_dir/ | wc -l
 grep -rnw 'SOEF Network Connection Error' output_dir/ |  wc -l
 grep -rnw 'SOEF Server Bad Response Error' output_dir/ |  wc -l
+gret -rnw ' Connection reset by peer' output_dir/ |  wc -l
```
Comment (marcofavorito) [inline examples/tac_deploy/README.md:98]:
```suggestion
grep -rnw ' Connection reset by peer' output_dir/ |  wc -l
```

File: packages/fetchai/connections/soef/connection.py [hunk 2]
```diff
@@ -1071,14 +1073,23 @@ async def _check_server_reachable(self) -> None:
...
+        reachable_check_count = 0
+        while reachable_check_count < self.connection_check_max_retries:
+            reachable_check_count += 1
+            try:
+                await self._check_server_reachable()
+                reachable_check_count = self.connection_check_max_retries
+            except Exception as e:  # pragma: nocover
+                if reachable_check_count == self.connection_check_max_retries:
+                    raise e
+                self.logger.debug(f"Exception during SOEF reachability check: {e}.")
```
Comment (marcofavorito) [inline packages/fetchai/connections/soef/connection.py:1092]: consider handling `SOEFNetworkConnectionError` explicitly, and print an appropriate error message.

Comment (marcofavorito) [reply packages/fetchai/connections/soef/connection.py:1092]: Something like:
```suggestion
            except SOEFNetworkConnectionError as e:
                if reachable_check_count == self.connection_check_max_retries:
                    self.logger.error("Max retries")
                    raise #TODO MaxRetriesError
                self.logger.debug(f"Exception during SOEF reachability check: {e}.")
```

**Review by solarw — APPROVED** [review]
LGTM

**Review by marcofavorito — APPROVED** [review]
LGTM

File: packages/fetchai/connections/soef/connection.py (hunk 2 above)
Comment (DavidMinarsch) [reply packages/fetchai/connections/soef/connection.py]: The problem is where missing multiple connection errors (not a full enumeration is done). I'd rather capture any exception until this is properly broken down.

**Pull request merged**